py-pyside2: update to 5.15.3 and add py310

abey79 opened the pull request:
#### Description

- updated pyside2 to 5.15.2.1
- added py310
- added a patch related to _Py_Mangle missing in the 5.15.2.1 source archive
- suppressed `ifconfig` directives from shiboken2 documentation which were generating an error

Note:
- this would be a major step to migrate #14251 to python 3.10
- there is a possible interaction with #14204 (?), which would probably best be merged before

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 12.2.1 21D62 arm64
Xcode 13.2.1 13C100

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

abey79 commented:
There seem to be some error related to Python 2.7. Seeing that nothing depends on py27-pyside2, would it be acceptable to drop this subport?

abey79 commented:
Rebased to use qt 5.15.3 (#14204)

abey79 commented:
Again, build fails with Python 2.7, but appears to succeed otherwise (including Python 3.10).

@reneeotten I noticed you recently removed quite a few 2.7/3.6 sub-ports lately. Would it be an options of this package? There are no dependant for the py27 nor py3{6,7,8} sub-ports.

@chrstphrchvz Considering all your work on Qt, any advice on the question?

chrstphrchvz commented on python/py-pyside2/Portfile:
```diff
@@ -4,7 +4,7 @@ PortSystem              1.0
 name                    py-pyside2
-version                 5.15.2
+version                 5.15.2.1
 revision                1
```
Revision should be reset to 0 (although I personally object to having to do so: if you forgot to reset the revision before building it, you and any other testers for a PR now get to reinstall and possibly rebuild it all over again, which I believe is pointless for ports that take a long time to build).

reneeotten replied:
this should indeed be set to `0`

reneeotten commented:
> Again, build fails with Python 2.7, but appears to succeed otherwise (including Python 3.10).
> 
> @reneeotten I noticed you recently removed quite a few 2.7/3.6 sub-ports lately. Would it be an options of this package? There are no dependant for the py27 nor py3{6,7,8} sub-ports.

I have no objections for removing EOL subports as long as there are no dependents (and the maintainers agree). However, the port `py-pyface` has a `pyside2` variant that depends on `pyXY-pyside2`, so if you want to remove the PY27/PY36 subports here you'll need to remove the py36 subport there as well and go down the dependency tree to remove all py36 subports so that things don't get into a broken state.
For example:
```
 % port echo depends:py36-pyface
py36-traitsui

% port echo depends:py36-traitsui
py36-apptools
py36-mayavi

% port echo depends:py36-apptools
py36-envisage
py36-mayavi

etcetera
```

chrstphrchvz commented:
> @chrstphrchvz Considering all your work on Qt, any advice on the question?

I really cannot answer authoritatively. PySide2 apparently intends to remain compatible with Python 2, given the availability of Python2.7 wheels for 5.15.2.1 on PyPI, so the py27-pyside2 build failure is likely unexpected. But I also have no desire to encourage others to continue using/offering ports for EOL Python.

abey79 commented:
> so what's the failure with Python 2.7 (it seems to me the CI is still busy and I don't really see that failure yet, or am I missing something?)

Here is what I could find:

```
2022-03-23T14:42:36.8501740Z cd /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/pyside2_build/py2.7-qt5.15.3-64bit-release/shiboken2/libshiboken && /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/compwrap/cxx/usr/bin/clang++ -DBUILD_LIBSHIBOKEN -DHAVE_NUMPY -DNDEBUG -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -I/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken -I/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/pyside2_build/py2.7-qt5.15.3-64bit-release/shiboken2/libshiboken -I/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/numpy/core/include -I/opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -arch x86_64 -stdlib=libc++ -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -Wall -fvisibility=hidden -Wno-strict-aliasing -D QT_NO_CAST_FROM_ASCII -D QT_NO_CAST_TO_ASCII -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.15 -fPIC   -fPIC -std=gnu++11 -MD -MT libshiboken/CMakeFiles/libshiboken.dir/voidptr.cpp.o -MF CMakeFiles/libshiboken.dir/voidptr.cpp.o.d -o CMakeFiles/libshiboken.dir/voidptr.cpp.o -c /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken/voidptr.cpp
2022-03-23T14:42:37.2641270Z /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken/pep384impl.cpp:757:19: error: use of undeclared identifier 'PyUnicode_GET_LENGTH'
2022-03-23T14:42:37.2672370Z     size_t plen = PyUnicode_GET_LENGTH(privateobj.object());
2022-03-23T14:42:37.2689680Z                   ^
2022-03-23T14:42:37.2707690Z /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken/pep384impl.cpp:760:12: error: use of undeclared identifier 'PyUnicode_READ_CHAR'
2022-03-23T14:42:37.2727750Z     while (PyUnicode_READ_CHAR(privateobj.object(), ipriv) == '_')
2022-03-23T14:42:37.2735100Z            ^
2022-03-23T14:42:37.2736450Z /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken/pep384impl.cpp:781:9: error: no matching function for call to 'PyUnicodeUCS2_AsWideChar'
2022-03-23T14:42:37.2761480Z     if (PyUnicode_AsWideChar(privateobj, resbuf + 1, ipriv + plen) < 0)
2022-03-23T14:42:37.2775650Z         ^~~~~~~~~~~~~~~~~~~~
2022-03-23T14:42:37.2837380Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:159:31: note: expanded from macro 'PyUnicode_AsWideChar'
2022-03-23T14:42:37.2863680Z # define PyUnicode_AsWideChar PyUnicodeUCS2_AsWideChar
2022-03-23T14:42:37.2897010Z                               ^~~~~~~~~~~~~~~~~~~~~~~~
2022-03-23T14:42:37.2939590Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:591:24: note: candidate function not viable: no known conversion from 'Shiboken::AutoDecRef' to 'PyUnicodeObject *' for 1st argument
2022-03-23T14:42:37.3001490Z PyAPI_FUNC(Py_ssize_t) PyUnicode_AsWideChar(
2022-03-23T14:42:37.3002860Z                        ^
2022-03-23T14:42:37.3019840Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:159:31: note: expanded from macro 'PyUnicode_AsWideChar'
2022-03-23T14:42:37.3045550Z # define PyUnicode_AsWideChar PyUnicodeUCS2_AsWideChar
2022-03-23T14:42:37.3064590Z                               ^
2022-03-23T14:42:37.3112120Z /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pyside2/py27-pyside2/work/pyside-setup-opensource-src-5.15.2/sources/shiboken2/libshiboken/pep384impl.cpp:783:9: error: no matching function for call to 'PyUnicodeUCS2_AsWideChar'
2022-03-23T14:42:37.3157830Z     if (PyUnicode_AsWideChar(name, resbuf + ipriv + plen + 1, nlen) < 0)
2022-03-23T14:42:37.3225680Z         ^~~~~~~~~~~~~~~~~~~~
2022-03-23T14:42:37.3232750Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:159:31: note: expanded from macro 'PyUnicode_AsWideChar'
2022-03-23T14:42:37.3248850Z # define PyUnicode_AsWideChar PyUnicodeUCS2_AsWideChar
2022-03-23T14:42:37.3262120Z                               ^~~~~~~~~~~~~~~~~~~~~~~~
2022-03-23T14:42:37.3299580Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:591:24: note: candidate function not viable: no known conversion from 'PyObject *' (aka '_object *') to 'PyUnicodeObject *' for 1st argument
2022-03-23T14:42:37.3305360Z PyAPI_FUNC(Py_ssize_t) PyUnicode_AsWideChar(
2022-03-23T14:42:37.3330690Z                        ^
2022-03-23T14:42:37.3354590Z /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/unicodeobject.h:159:31: note: expanded from macro 'PyUnicode_AsWideChar'
2022-03-23T14:42:37.3373650Z # define PyUnicode_AsWideChar PyUnicodeUCS2_AsWideChar
2022-03-23T14:42:37.3416650Z                               ^
2022-03-23T14:42:37.3440830Z 4 errors generated.
```

It look like misuse of some CPython API (of which I admit knowing next to nothing). It's somewhat reminiscent of the error I got on 3.10 and for which I needed to pull a patch from some forum post (URL in the Portfile). It looks like they really rushed this release (originally motivated by adding 3.10 compatibility). Whatever 3.10 binaries are on PyPI, they cannot be built from the unmodified source. It may be the case for the 2.7 binaries as well.

> However, the port py-pyface has a pyside2 variant that depends on pyXY-pyside2, so if you want to remove the PY27/PY36 subports here you'll need to remove the py36 subport there as well and go down the dependency tree to remove all py36 subports so that things don't get into a broken state.

Oh, I missed that, thanks for pointing out. Well, py36-pyside2 builds alright, so we can leave it here and just remove 2.7.

Noob side question: how do I find *all* XY dependants, including variants?

abey79 commented:
I have fixed the revision and removed py27 to see if the CI is ok with the rest. Of course, if there is disagreement about dropping py27, I'm happy to put it back.

abey79 commented:
IMHO, this is ready for merging **iff** dropping 2.7 is acceptable @reneeotten @MarcusCalhoun-Lopez @pmetzger

reneeotten commented on python/py-pyside2/Portfile:
```diff
@@ -4,8 +4,8 @@ PortSystem              1.0
 name                    py-pyside2
-version                 5.15.2
-revision                1
+version                 5.15.2.1
```
latest upstream release is 5.15.3 - can you please update to that version?

abey79 replied:
done

reneeotten commented on python/py-pyside2/Portfile:
```diff
@@ -16,19 +16,25 @@ homepage                https://wiki.qt.io/Qt_for_Python
 license                 {GPL-2 GPL-3+ LGPL-3}
 master_sites            https://download.qt.io/official_releases/QtForPython/pyside2/PySide2-${version}-src
 distname                pyside-setup-opensource-src-${version}
-use_xz                  yes
```
this should stay since that latest version again provides a `.tar.xz` file and these archives are smaller

abey79 replied:
👍🏻

reneeotten commented on python/py-pyside2/Portfile:
```diff
@@ -16,19 +16,25 @@ homepage                https://wiki.qt.io/Qt_for_Python
+
+    # fix error with shiboken2 doc building
+    patchfiles-append   patch-shiboken2-docs.diff
+
```
perhaps these patches are not needed anymore?

abey79 replied:
Nope, build still fails when either patch is removed

reneeotten commented:
ping @abey79

abey79 commented:
@reneeotten Yeah, sorry for the radio silence. I've had a few busy days, will get to it ASAP.

pmetzger commented:
I'd remove 2.7 support FWIW. Sorry I was absent.

abey79 commented:
Sorry for the delay, I just pushed the updated Portfile.

sumirati commented:
Apologies if I did not follow the etiquette - this is my first try on Github merge request workflows.
I opened #14572 as I had to do local changes to allow py-pyside2 to compile against this nice work.

This might be related to the clang bug, that has also shown up in other places (ie #14470)
